gui/liguidsoap.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH LIGUIDSOAP 1 "march 16, 2007"
.\" Please adjust this date whenever revising the manpage.
.\"
.SH NAME
liguidsoap \- Control GUI for liquidsoap

.SH DESCRIPTION
This manual page documents briefly the
.B liguidsoap
control GUI for
.BR liquidsoap (1).

.SH OPTIONS
.TP
.B \-h, \-\-host
Host where liquidsoap is running
.TP
.B \-p, \-\-port
Port where liquidsoap is listening

.SH SEE ALSO
.BR liquidsoap (1).

.SH AUTHOR
.B liguidsoap
was written by The Savonet Team <savonet-users@lists.sf.net>.
.PP
This manual page was written by Romain Beauxis <toots@rastageeks.org>,
for the Debian project  Permission is granted to copy, distribute and/or
modify this document under the terms of the GNU General Public
License, Version 2 any later version published by the Free Software Foundation.


CHANGES
0.3.8 (30-07-2008)
Bugs fixed:
 - Vorbis mono output is now working
 - Fixed parameter parameter description
   in the documentation
 - Propagate new delay in add_timeout
 - Fixed inter-thread mutex lock/unlock in playlist.ml
 - Fixed "next" playlist command
 - Fixed race conditions in request_source.ml feeding task
 - cross/smartcross: raise the default for inhibition as 
   setting it to exactly duration is not enough
 - Don't fail when $HOME is not set
 - Fixed metadata update in input.harbor with icecast2 source protocol
 - Fixed shutdown function. Fixes #153
 - Fixed input.oss. Liquidsoap now works with OSS4 ! Fixes #158
New:
 - Added a hook to execute a function when playlist.once ends
 - Enhanced smart_crossfade
 - Added string.case and string.capitalize
 - New "exec_at" operator, to execute a function depending on a given predicate
 - Added script example in the documentation to listen to radio Nova and get 
   the metadata from a web page.
 - Changed parameters name in fallback.skip to reflect who are the fallback
   and input source.
 - Added a dump file parameter to input.harbor, for debugging purpose.
 - Added an auth function parameter to input.harbor for custom 
   authentifications. Fixes #157
 - Added "primary_queue" and "secondary_queue" commands to request.queue and
   request.equeue sources. Also set the metadata "queue" to either "primary"
   or "secondary" for each request in the queue. Documented it too.
 - Insert latest metadata for a node of a switch source when switching 
   back to it in a middle of a track.
 - Added a 'conservative' parameter to cross, smilar to the one in smartcross.
Internal:
 - Enhanced liqi documentation parser to build the website.

0.3.7 (03-06-2008)
Bugs fixed:
 - Now works on FreeBSD and hopefully other unices that are stricter than
   Linux about Mutex usage.
 - input.http() now has a bind_address parameter.
 - Harbor socket now has a timeout for lost connections.
 - smartcross() is now more compliant with the inter-sources protocol,
   fixes several "get frame didn't change the buffer" bugs.
 - Ogg packeting bugs.
 - Buffering policy in input.http/harbor().
 - No "." in IDs and labels.
 - Resources: FD leaks, useless threads (threads leaks?) in input.http().
 - fade.out() used to run into infinite loops when the delay was 0.
New:
 - New documentation system and website.
 - Self-documenting server with a more helpful "help" command.
 - Moved to duppy: less threads, lighter load, and an configurable scheduler.
 - Moved to Taglib for more reliable access to MP3 metadata.
 - MIME types, notably for playlists and MP3 files.
 - New Jack support. The old one has been renamed to in/output.jack.legacy().
 - Harbor: per-mount passwords and the stop command to kick a source client.
 - Official Last.FM client.
 - Metadata is no more punctual but interval-based, which suppresses some 
   surprising behaviours.
 - Perfected daemon behaviour.
 - All output.file.*() now have the features that used to be only in 
   output.file.vorbis(), notable re-opening. Added %w to the strftime-like 
   format codes allowed in their filename parameter.
 - Add clear_metadata() and map_metadata(). Now, rewrite_metadata() is a simple
   specialization of map_metadata(), written in utils.liq.
 - Dynamic amplification factor in amplify(), e.g. useful for replay gain.
 - Lots of new functions in the scripting API: for lists, requests, system 
   interaction, shutdown, command-line parsing, scripted server commands, etc.
As always:
 - code cleanup, style, etc.

0.3.6 (17-12-2007)
Bugfix release:
 - Close Http socket
 - Add http socket timeout
 - Close playlist file after reading its content
 - Fix http redirect support for lastfm files
 - Fix file open leak in camomile support
 - Fix playlist uri ending with "/"

0.3.5 (08-11-2007)
Bugfix release:
  - Fixed #57: scpls and mpegurl playlist parsing
  - Fixed #46: Late cross-scripts bindings

0.3.4 (25-09-2007)
Notation: "-" stands for a change, "+" for an addition.
* Language
  - Support for polymorphism, subtyping and basic ad-hoc polymorphism,
    which allows a much simpler API, notably for maths and serialization.
  + Added interactive_*() for mutable values.
  - The right syntax for settings is now set("var", value) and can be used
    anywhere in the scripts.
  - The volume parameters of most operators are now in dB.
  - Many builtin functions added.
  - Nicer type error messages.
* Sources
  + Added input.lastfm() to relay last.fm streams.
  + Added input.harbor() to received Icecast2 source streams.
  + Added noise() to generate white noise
  - Reimplemented playlist support, added various xml and text formats.
  - Added mpd protocol to find files using mpd.
* Operators
  + New effects: compress(), flanger(), pan().
  + New filters: filter.fir.*(), filter.iir.*(), filter.biquad.*(), comb().
  + Added support for LADSPA effects.
  + Added eat_blank() to remove blanks.
* Outputs
  - Added non-default restart option for output.icecast.*().
  - Added the possibility to tweak some settings at runtime.
  - Split output.icecast.vorbis() into output.icecast.vorbis.*() to distinguish
    between encoding modes -- and similarly for output.file.vorbis and mp3.
  - Better handling of Icecast disconnections.
* IO
  + Added portaudio support.
  - Jack support is now somewhat working.
* As usual, lots of bug fixes, careful polishing & much more...

0.3.3 (06-06-2007)
* Major cleanup of the core stream representation; moved to float arrays,
  removing several back-and-forth conversions and enhancing the perfs a lot;
  reviewed all sources and operators, made many minor enhancements btw.
* Lots of sound processing operators: compand, compress, normalize,
  pitch, bpm, soundtouch, saw, square, etc. Add more shapes to fade.*().
* New track processing operators: insert_metadata, on_track.
* Smart cross: allows to select a transition based on the volumes around the 
  end-of-track.
* Support for AAC encoding/decoding.

gui/liguidsoap.py
#!/usr/bin/env python

import pygtk
pygtk.require('2.0')
import gtk
import gobject

import threading, socket, sys, os, time, re
import getopt

from client import LiqClient
from widgets import View

from queue import LiqQueue
from editable import LiqEditable
from playlist import LiqPlaylist
from mix import LiqMix
from output import LiqOutput

class LiqGui(gtk.Notebook):
  def __init__(self,host='localhost',port=1234):
    gtk.Notebook.__init__(self)
    self.set_show_tabs(False)

    self.host=host
    self.port=port
    self.tel = LiqClient(host,port)
    self.list = View([['type','80'],['name','100']],[])
    self.list.connect('row_activated',self.row_activated)
    box = gtk.HBox()
    box.pack_start(self.list,False,False)
    lbl = gtk.Label()
    lbl.set_markup("""
<b>The Savonet Team is happy to welcome you in liguidsoap.</b>

On the left is a list of controllable nodes. Double-click a row to open the controller.

The <b>output</b> nodes are those which do something with your stream.
Some output directly to your speakers, some save your stream to a file, and 
finally, some output to an icecast server for webradio.

A <b>mixer</b> is a mixing table. It has some input sources, and allows you 
to select which one you want to play, tune the volume, etc.

<b>editable</b> and <b>queue</b> nodes are interactive playlists, in which
you can add files using drag-n-drop or builtin file browser. The editable is 
more powerful, allows you to delete, insert at any place, not only at the 
bottom.

Finally, a <b>playlist</b> is a non-interactive playlist, which list and 
behaviour is hard-coded in the liquidsoap script. You can just see what will be 
played next. (Actually, more could be coming soon, like changing random mode 
and playlist file...)

Liquidsoap and liguidsoap are copyright (c) 2003,2004,2005 Savonet Team.
This is free software, released under the terms of GPL version 2 or higher.
""")
    box.pack_start(lbl)
    self.append_page(box,gtk.Label('list'))
    box.show() ; lbl.show()
    self.list.show()
    self.show()
    gobject.timeout_add(1000,self.update)

  def update(self):
    def hashof(s):
      m = re.search('(\S+)\s:\s(\S+)',s)
      return { 'name' : m.group(1), 'type' : m.group(2) }
    list = [ hashof(s) for s in
             filter(lambda x: x!='',
                    re.compile('\n').split(self.tel.command('list'))) ]
    self.list.setModel(list)

  def row_activated(self,view,path,column):
    type = view.get_model().get_value(view.get_model().get_iter(path),0)
    name = view.get_model().get_value(view.get_model().get_iter(path),1)
    b = gtk.HBox()
    l = gtk.Label(name+' ')
    c = gtk.Button('x') # TODO stock close
    c.n = self.get_n_pages()
    b.pack_start(l)
    b.pack_start(c)
    c.show() ; l.show()
    def clicked(w):
      self.remove_page(w.n) # TODO cleanup contents (especially telnet)
    c.connect('clicked',clicked)
    self.set_show_tabs(True)
    self.append_page(
        {'queue'   : lambda x: LiqQueue(self.host,self.port,x),
         'editable': lambda x: LiqEditable(self.host,self.port,x),
         'playlist': lambda x: LiqPlaylist(self.host,self.port,x),
         'mixer'   : lambda x: LiqMix(self.host,self.port,x),
         'output.icecast': lambda x: LiqOutput(self.host,self.port,x),
         'output.ao'     : lambda x: LiqOutput(self.host,self.port,x),
         'output.alsa'   : lambda x: LiqOutput(self.host,self.port,x),
         'output.ogg'    : lambda x: LiqOutput(self.host,self.port,x),
         'output.wav'    : lambda x: LiqOutput(self.host,self.port,x)
        }[type](name),b)

# Now is the big deal: how to start liguidsoap,
# configuration dialog and automatic liquidsoap launch.

# liquidsoap runs liquidsoap with a fixed script
# a few parameteres are available
def liquidsoap(
    host='localhost',port=1234,mount='emission.ogg',
    backup=''):
  if backup=='':
    addbackup=""
  else:
    addbackup=';"backup"'

  os.system("""cat > /tmp/liguidsoap.liq <<__EOL__
set log.dir = "/tmp"
set log.stdout = true
set telnet = true

bg = request.equeue (id="bed")
music = request.equeue (id="music")
sfx = request.equeue (id="sfx")
# Disabling input.alsa until it is more stable..
# Currently its initialization can freeze liquidsoap,
# and it won't run correctly on most systems.
mixer = mix(id="mixer",[
%ifdef input.alsa
                         input.alsa(id="microphone"),
%endif
                         music,bg,sfx])

# This is output.alsa if available, ao or dummy otherwise.
output.prefered(id="speaker",mixer)
output.icecast(
  id="broadcast",
  host="%s",port=%d,mount="%s",bitrate=64,start=false,mixer)
output.ogg(id="backup",start=false,"%s",mixer)
""" % (host, port, mount, backup))
  pid = os.fork()
  if pid==0:
    os.execlp("liquidsoap","liquidsoap","/tmp/liguidsoap.liq")
  else:
    print "Running liquidsoap..."
    return pid

# liguidsoap is the toplevel call, starts everything
def liguidsoap():
  try:
    opts, args = getopt.gnu_getopt(sys.argv[1:],"h:p:",['host=','port='])
  except:
    print 'Usage: %s [OPTIONS]...' % sys.argv[0]
    print 'Options are:'
    print '  -h,--host=HOST'
    print '  -p,--port=PORT'
    sys.exit()

  host=None
  port=1234
  for o , a in opts:
    if o in ('-p', '--port'):
      port=int(a)
    if o in ('-h', '--host'):
      host=a

  ehost=eport=erun=dialog=None
  icehost=iceport=icemount=backup=None

  def exit(pid):
    if pid!=None:
      os.kill(pid,15)
      os.waitpid(pid,0)
    gtk.main_quit()

  # This startup function can be used to start the GUI directly
  # or after user input in dialog
  def start(response=None):
    # Dialog stuff
    liquid_pid=None
    if response!=None:
      if response!=gtk.RESPONSE_ACCEPT:
        sys.exit()
      if erun.get_active():
        host,port = 'localhost',1234
        liquid_pid=liquidsoap(
            host=icehost.get_text(),
            port=iceport.get_value(),
            mount=icemount.get_text(),
            backup=backup.get_text())
        print "Waiting for liquidsoap to start..."
        time.sleep(3)
      else:
        host=ehost.get_text()
        port=int(eport.get_value())
      dialog.destroy()

    # Startup
    win = gtk.Window()
    win.set_border_width(10)
    win.connect("delete_event", lambda w,e: False)
    win.connect("destroy", lambda osb: exit(liquid_pid))
    win.set_title('Liquidsoap on '+host+':'+str(port))
    win.resize(700,300)
    try:
      win.add(LiqGui(host,port))
    except socket.error, x:
      error = gtk.MessageDialog(message_format=
          ("Couln't connect to "+host+' on port '+str(port)+'!'))
      # TODO there's a strange cursor in the message
      error.connect("response",gtk.main_quit)
      error.show()
      return
    win.show()

  # If no host has been given on command-line, use the dialog
  if host==None:
    dialog = gtk.Dialog(
        "LiGUIdsoap",None,0,
        (gtk.STOCK_CANCEL, gtk.RESPONSE_CANCEL,
          gtk.STOCK_OK, gtk.RESPONSE_ACCEPT,))
    dialog.connect("response",lambda d,id: start(id))

    table = gtk.Table(2,2)
    table.show()
    dialog.vbox.pack_start(table)

    ehost = gtk.Entry()
    ehost.set_text('localhost')
    ehost.show()
    lbl = gtk.Label("Host: ")
    lbl.show()
    table.attach(lbl,0,1,0,1)
    table.attach(ehost,1,2,0,1)

    eport = gtk.SpinButton()
    eport.set_range(0,99999)
    eport.set_increments(1,10)
    eport.set_value(port)
    eport.show()
    lbl = gtk.Label("Port: ")
    lbl.show()
    table.attach(lbl,0,1,1,2)
    table.attach(eport,1,2,1,2)

    # Settings for automatic liquidsoap run
    erunconf = gtk.Table(2,6)
    lbl = gtk.Label("Icecast settings")
    erunconf.attach(lbl,0,2,0,1) ; lbl.show()

    icehost = gtk.Entry()
    icehost.set_text('localhost')
    lbl = gtk.Label("Host: ")
    erunconf.attach(lbl,0,1,1,2)
    erunconf.attach(icehost,1,2,1,2)
    lbl.show() ; icehost.show()

    iceport = gtk.SpinButton()
    iceport.set_range(0,99999)
    iceport.set_increments(1,10)
    iceport.set_value(8000)
    iceport.show()
    lbl = gtk.Label("Port: ")
    lbl.show()
    erunconf.attach(lbl,0,1,2,3)
    erunconf.attach(iceport,1,2,2,3)

    icemount = gtk.Entry()
    icemount.set_text('emission.ogg')
    lbl = gtk.Label("Mount: ")
    erunconf.attach(lbl,0,1,3,4)
    erunconf.attach(icemount,1,2,3,4)
    lbl.show() ; icemount.show()

    lbl = gtk.Label("Local backup OGG file")
    erunconf.attach(lbl,0,2,4,5)
    backup = gtk.Entry()
    backup.set_text('/tmp/emission.ogg')
    erunconf.attach(backup,0,2,5,6)
    backup.show() ; lbl.show()

    erun = gtk.CheckButton("Run liquidsoap automatically.")
    erun.show()
    dialog.vbox.pack_start(erun)
    dialog.vbox.pack_start(erunconf)
    def erunclicked(w):
      if erun.get_active():
        erunconf.show()
      else:
        erunconf.hide()
    erun.connect("clicked", erunclicked)

    dialog.show()

  # Otherwise, start directly
  else:
    start()

  gtk.main()

if __name__=='__main__':
  liguidsoap()
